op-program/client/l2/engineapi/l2_engine_api.go

```diff
@@ -177,6 +178,82 @@ func (ea *L2EngineAPI) endBlock() (*types.Block, error) {
 }
 
 func (ea *L2EngineAPI) GetPayloadV1(ctx context.Context, payloadId eth.PayloadID) (*eth.ExecutionPayload, error) {
+	return ea.getPayload(ctx, payloadId)
+}
+
+func (ea *L2EngineAPI) GetPayloadV2(ctx context.Context, payloadId eth.PayloadID) (*eth.ExecutionPayloadEnvelope, error) {
+	payload, err := ea.getPayload(ctx, payloadId)
+	return &eth.ExecutionPayloadEnvelope{ExecutionPayload: payload}, err
+}
+
+func (ea *L2EngineAPI) config() *params.ChainConfig {
+	return ea.backend.Config()
+}
+
+func (ea *L2EngineAPI) ForkchoiceUpdatedV1(ctx context.Context, state *eth.ForkchoiceState, attr *eth.PayloadAttributes) (*eth.ForkchoiceUpdatedResult, error) {
+	if attr != nil {
+		if attr.Withdrawals != nil {
+			return STATUS_INVALID, engine.InvalidParams.With(errors.New("withdrawals not supported in V1"))
+		}
+		if ea.config().IsShanghai(ea.config().LondonBlock, uint64(attr.Timestamp)) {
```

Comment:
Is there any easy way to get the current block? As far as I can tell, the mix between the shanghai block & the london block will mean that this is always true or always false.

Reply:
I believe it should work as expected, the underlying `IsShanghai` [method](https://github.com/ethereum-optimism/op-geth/blob/optimism/params/config.go#L582-L585) uses the second arg, in this case `attr.Timestamp` to detect if it's Shanghai or not. The first arg (the london block) is used just to see if we've forked past London.

For context, this code is ported from [op-geth](https://github.com/ethereum-optimism/op-geth/blob/optimism/eth/catalyst/api.go#L175-L182), my intention was to try and keep it as similar as possible. However happy to try and resolve the closest block to `attr.Timestamp` if you think it's worth it.

Reply:
My bad, I didn't realize that geth requires timestamp forks to be post london.